walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,60 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.List;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    //Approach: start form start of day, and add available timeranges as we traverse through the events' timeranges
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        Collection<TimeRange> validTimeRanges = new ArrayList();
+        long duration = request.getDuration();
+        int start = 0;
+        int end = 0;
+        int lastEnd = 0;
+
+        // Get all the event time ranges and sort them by the start time. 
+        // Only add to the list if at least one attendee from the event is 
+        //a required attendee in the meeting request. (Otherwise, the event is irrelevant.)
+        List<TimeRange> eventTimeRanges = new ArrayList();
+        for (Event event: events) {
+            for (String attendee: event.getAttendees()){
+                if (request.getAttendees().contains(attendee)){
+                    eventTimeRanges.add(event.getWhen());
+                    break;
+                }
+            }
+        }
```

Comment:
could this be expressed using Java 8's stream API?

something like (I haven't tested):

```suggestion
         events.filter(
         e -> event.getAttendees().stream().filter(
             a -> request.getAttendees().contains(attendee))
         .getAny().isPresent())
         .map(e -> e.getWhen())
         .collect(Collectors.ToList());
```

Reply:
Thanks for sharing! Would this result in longer runtime though? Because it would check every attendee in each event to filter them, and then filter the events?

Reply:
the Java stream API will is allowed to do this optimization when you add getAny()

Reply:
Also, I found that you can write the inner matcher with a anyMatch():

event.getAttendees().stream().anymatch(a -> request.getAttendees().contains(attendee))

Reply:
Done, thank you!